Assets/UI/worldinput_CQUI.lua

```diff
@@ -0,0 +1,768 @@
+    return true;
+  end
+
+  -- Allow other handlers of KeyDown events to capture input, if any exist after World Input
+  return false;
+end
+
+-- ===========================================================================
+function DefaultKeyUpHandler( uiKey:number )
+  print_debug("** Function Entry: DefaultKeyUpHandler (CQUI Hook) uiKey: "..tostring(uiKey));
+
+  -- If set to Standard, just let the game do its thing
+  if CQUI_hotkeyMode == CQUI_HOTKEYMODE_STANDARD then
+    if uiKey == Keys.VK_SHIFT then
+      CQUI_isShiftDown = false;
+    end
+  
+    if uiKey == Keys.VK_ALT then
+      CQUI_isAltDown = false;
+    end
+
+    return BASE_CQUI_DefaultKeyUpHandler(uiKey);
+  end
+
+  local keyPanChanged  :boolean = false;
+  local cquiHandledKey :boolean = false;
+
+  local selectedUnit = UI.GetHeadSelectedUnit();
+  local unitType = nil;
+  local formationClass = nil;
+  if (selectedUnit) then
+    unitType = GameInfo.Units[selectedUnit:GetUnitType()].UnitType;
+    formationClass = GameInfo.Units[selectedUnit:GetUnitType()].FormationClass;
+  end
+
+  --CQUI Keybinds
+  if CQUI_hotkeyMode == CQUI_HOTKEYMODE_ENHANCED then
+    if (uiKey == Keys.W) then
+      m_isUPpressed = false;
+      keyPanChanged = true;
+    end
+
+    if (uiKey == Keys.D) then
+      m_isRIGHTpressed = false;
+      keyPanChanged = true;
+    end
+
+    if (uiKey == Keys.S and CQUI_isAltDown == false) then
+      m_isDOWNpressed = false;
+      keyPanChanged = true;
+    end
+
+    if (uiKey == Keys.A) then
+      m_isLEFTpressed = false;
+      keyPanChanged = true;
+    end
+
+    if (uiKey == Keys.E) then
+      if (CQUI_isAltDown == true) then
+        -- Behave as if just "E" was pressed, send the unit exploring
+        -- Calling this using UnitOperationTypes doesn't work as Automate_Explore hash does not appear among the GameInfo.OperationTypes
+        UnitManager.RequestOperation(UI.GetHeadSelectedUnit(), GameInfo.UnitOperations["UNITOPERATION_AUTOMATE_EXPLORE"].Hash);
+        cquiHandledKey = true;
+      elseif (CQUI_cityview) then
+        LuaEvents.CQUI_GoNextCity();
+        cquiHandledKey = true;
+      else
+        UI.SelectNextReadyUnit();
+        cquiHandledKey = true;
+      end
+    end
+
+    if (uiKey == Keys.Q) then
+      if (CQUI_isAltDown == true and unitType == "UNIT_BUILDER" ) then
+        -- Do nothing, this case is handled later on in this function
+      elseif (CQUI_cityview) then
+        LuaEvents.CQUI_GoPrevCity();
+        cquiHandledKey = true;
+      else
+        UI.SelectPrevReadyUnit();
+        cquiHandledKey = true;
+      end
+    end
+
+    if (uiKey == Keys.VK_SHIFT and ContextPtr:LookUpControl("/InGame/TechTree"):IsHidden() and ContextPtr:LookUpControl("/InGame/CivicsTree"):IsHidden()) then
+      cquiHandledKey = true;
+      if (CQUI_cityview) then
+        UI.SetInterfaceMode(InterfaceModeTypes.SELECTION);
+        UI.SelectNextReadyUnit();
+      else
+        LuaEvents.CQUI_GoNextCity();
+      end
+    end
```

Comment:
This is just an observation, no code change required for this pull.
The unmodded game allows pressing Shift+Enter to force end turn. This means you don't have to cycle through all your units and tell them to skip turn. Very helpful for late game large army games. I think having this section here results in the view jumping around when the Shift key is released after the forced end turn. I'll file an issue. Just wanted to point at code ask what your thoughts were regarding the behaviour.

Reply:
I am not a fan of the shift-to-show-city thing, but I also don't use the keyboard shortcuts myself.  I know the shift key was not showing the city with the Shift-P (with last night's changes) but didn't try the Shift-Enter

Reply:
In general I think modifier keys (Alt, Shift, Ctrl) should be just that, modifiers. They should not be used for any specific action by themselves. Anyways, I'm just biased against this hotkey. I've seen weird behaviour when doing Shift+Enter while running CQUI. I'll try to right up some stuff for why I think this one is a bad idea and under what conditions I see weird behaviour when I file the issue later.

Reply:
I agree - I'll take it out and add it to that issue for the airbases and stuff, can figure it out later